lib/api/api_auth.go

```diff
@@ -37,15 +37,66 @@ func emitLoginAttempt(success bool, username, address string, evLogger events.Lo
 	}
 }
 
-func basicAuthAndSessionMiddleware(cookieName string, guiCfg config.GUIConfiguration, ldapCfg config.LDAPConfiguration, next http.Handler, evLogger events.Logger) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+func unauthorized(w http.ResponseWriter) {
+	time.Sleep(time.Duration(rand.Intn(100)+100) * time.Millisecond)
+	w.Header().Set("WWW-Authenticate", "Basic realm=\"Authorization Required\"")
+	http.Error(w, "Not Authorized", http.StatusUnauthorized)
+}
+
+func forbidden(w http.ResponseWriter) {
+	time.Sleep(time.Duration(rand.Intn(100)+100) * time.Millisecond)
+	http.Error(w, "Forbidden", http.StatusForbidden)
+}
+
+func equalsAny(s string, values []string) bool {
+	for _, value := range values {
+		if s == value {
+			return true
+		}
+	}
+	return false
+}
+
+func hasAnyPrefix(s string, prefixes []string) bool {
+	for _, prefix := range prefixes {
+		if strings.HasPrefix(s, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func noAuthPaths() []string {
+	return []string{
+		"/",
+		"/index.html",
+		"/modal.html",
+	}
+}
+
+func noAuthPrefixes() []string {
+	return []string{
+		// Static assets
+		"/assets/",
+		"/syncthing/",
+		"/vendor/",
+		"/theme-assets/", // This leaks information from config, but probably not sensitive
+
+		// No-auth API endpoints
+		"/rest/noauth",
+	}
+}
```

Comment:
Can these be simple `var`s or even `const`, or why do we need a function call?

Reply:
I made them `const`s first, but as far as I know there is no way to make slices immutable in Go. Creating them anew on each use seems to be the only workaround. Since these are used for security decisions, I chose correctness (make it impossible to accidentally edit them) over performance here. I'll add a comment explaining this and a test enforcing it.